From: Sebastian Keller <skeller@gnome.org>
Date: Fri, 11 Apr 2025 14:28:35 +0200
Subject: cairoblur: Limit repeated blur clip size after scaling

The clip's width/height was being limited to 1 pixel for repeated blurs,
but that happened before scaling. So on 2x scaling this would get scaled
up to 2 pixels afterwards, unnecessarily doubling the number of pixels
to blurred for repeated blurs.

Part-of: <https://gitlab.gnome.org/GNOME/gtk/-/merge_requests/8461>
(cherry picked from commit 08a6c85fd9b9468030d2d1e114c6797cd68ddb38)

Origin: upstream gtk-4-18 after 4.18.4
---
 gsk/gskcairoblur.c | 7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

diff --git a/gsk/gskcairoblur.c b/gsk/gskcairoblur.c
index 339479f..909ad8f 100644
--- a/gsk/gskcairoblur.c
+++ b/gsk/gskcairoblur.c
@@ -360,6 +360,9 @@ gsk_cairo_blur_start_drawing (cairo_t         *cr,
   cairo_surface_get_device_scale (cairo_get_target (cr), &x_scale, &y_scale);
   apply_current_scale (cr, &x_scale, &y_scale);
 
+  clip_width = x_scale * (clip_width + (blur_x ? 2 * clip_radius : 0));
+  clip_height = y_scale * (clip_height + (blur_y ? 2 * clip_radius : 0));
+
   if (blur_flags & GSK_BLUR_REPEAT)
     {
       if (!blur_x)
@@ -371,8 +374,8 @@ gsk_cairo_blur_start_drawing (cairo_t         *cr,
   /* Create a larger surface to center the blur. */
   surface = cairo_surface_create_similar_image (cairo_get_target (cr),
                                                 CAIRO_FORMAT_A8,
-                                                x_scale * (clip_width + (blur_x ? 2 * clip_radius : 0)),
-                                                y_scale * (clip_height + (blur_y ? 2 * clip_radius : 0)));
+                                                clip_width,
+                                                clip_height);
   cairo_surface_set_device_scale (surface, x_scale, y_scale);
   cairo_surface_set_device_offset (surface,
                                     x_scale * ((blur_x ? clip_radius : 0) - clip_x1),
